crates/matrix-sdk-ui/src/timeline/threaded_events_loader.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{fmt::Formatter, sync::Mutex};
+
+use matrix_sdk::{
+    event_cache::{
+        paginator::{PaginationResult, PaginatorError},
+        PaginationToken,
+    },
+    room::{IncludeRelations, RelationsOptions},
+};
+use ruma::{api::Direction, events::relation::RelationType, OwnedEventId, UInt};
+
+use super::traits::RoomDataProvider;
+
+#[derive(Debug)]
+struct PaginationTokens {
+    previous: PaginationToken,
+    next: PaginationToken,
+}
+
+pub struct ThreadedEventsLoader<P: RoomDataProvider> {
+    room: P,
+    root_event_id: OwnedEventId,
+    tokens: Mutex<PaginationTokens>,
+}
+
+impl<P: RoomDataProvider> ThreadedEventsLoader<P> {
+    /// Create a new [`Paginator`], given a room implementation.
+    pub fn new(room: P, root_event_id: OwnedEventId) -> Self {
+        Self {
+            room,
+            root_event_id,
+            tokens: Mutex::new(PaginationTokens { previous: None.into(), next: None.into() }),
+        }
+    }
+
+    pub async fn paginate_backwards(
+        &self,
+        num_events: UInt,
+    ) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Backward, num_events).await
```

Comment:
Since we're not likely to implement forward pagination, let's inline `paginate` in there, so then we remove the `Direction` parameter, and likely simplify the code in `paginate` too!

Reply:
We will most likely have to implement forward pagination at some point for thread permalink support but I agree that it won't probably be here.